etcdserver/etcdhttp/client_security.go

```diff
@@ -88,23 +91,34 @@ func hasKeyPrefixAccess(sec *security.Store, r *http.Request, key string) bool {
 		log.Printf("security: Incorrect password for user: %s.", username)
 		return false
 	}
-	if user.User == "root" {
-		return true
-	}
 	writeAccess := r.Method != "GET" && r.Method != "HEAD"
 	for _, roleName := range user.Roles {
 		role, err := sec.GetRole(roleName)
 		if err != nil {
 			continue
 		}
-		if role.HasKeyAccess(key, writeAccess) {
-			return true
+		if recursive {
+			return role.HasRecursiveAccess(key, writeAccess)
 		}
+		return role.HasKeyAccess(key, writeAccess)
 	}
 	log.Printf("security: Invalid access for user %s on key %s.", username, key)
 	return false
 }
 
+func hasGuestAccess(sec *security.Store, r *http.Request, key string) bool {
+	writeAccess := r.Method != "GET" && r.Method != "HEAD"
+	role, err := sec.GetRole(security.GuestRoleName)
+	if err != nil {
+		return false
+	}
+	if role.HasKeyAccess(key, writeAccess) {
+		return true
+	}
+	log.Printf("security: Invalid access for unauthenticated user on key %s.", key)
```

Comment:
for consistency: in the hasRootAccess we say resource, in this message we say key.